src/ai/backend/storage/vast/__init__.py

```diff
@@ -0,0 +1,244 @@
+        if vast_quota_id is None:
+            raise QuotaScopeNotFoundError
+        try:
+            await self.api_client.modify_quota(
+                vast_quota_id,
+                soft_limit=config.limit_bytes,
+                hard_limit=config.limit_bytes,
+            )
+        except VastInvalidParameterError:
+            raise InvalidQuotaConfig
+        except VastUnknownError as e:
+            raise ExternalError(str(e))
+
+    async def describe_quota_scope(self, quota_scope_id: QuotaScopeID) -> Optional[QuotaUsage]:
+        qspath = self.mangle_qspath(quota_scope_id)
+        if not qspath.exists():
+            return None
+        if (vast_quota_id := await self._get_vast_quota_id(quota_scope_id)) is None:
+            return None
+        if (quota := await self.api_client.get_quota(vast_quota_id)) is None:
+            return None
+        return QuotaUsage(
+            used_bytes=quota.used_capacity,
+            limit_bytes=quota.hard_limit,
+        )
+
+    async def unset_quota(self, quota_scope_id: QuotaScopeID) -> None:
+        vast_quota_id = await self._get_vast_quota_id(quota_scope_id)
+        if vast_quota_id is None:
+            raise QuotaScopeNotFoundError
+        try:
+            await self.api_client.remove_quota(vast_quota_id)
+        except VastNotFoundError:
+            raise QuotaScopeNotFoundError
+        await self._rm_vast_quota_id(quota_scope_id)
+
+    async def delete_quota_scope(self, quota_scope_id: QuotaScopeID) -> None:
+        await self.unset_quota(quota_scope_id)
+        qspath = self.mangle_qspath(quota_scope_id)
+        await aiofiles.os.rmdir(qspath)
+
+
+class VastVolume(BaseVolume):
+    api_client: VastAPIClient
+
+    name = "vast"
+
+    def __init__(
+        self,
+        local_config: Mapping[str, Any],
+        mount_path: Path,
+        *,
+        etcd: AsyncEtcd,
+        event_dispathcer: EventDispatcher,
+        event_producer: EventProducer,
+        options: Optional[Mapping[str, Any]] = None,
+    ) -> None:
+        super().__init__(
+            local_config,
+            mount_path,
+            etcd=etcd,
+            options=options,
+            event_dispathcer=event_dispathcer,
+            event_producer=event_producer,
+        )
+        ssl_verify = self.config.get("vast_verify_ssl", False)
+        self.api_client = VastAPIClient(
+            self.config["vast_endpoint"],
+            self.config["vast_username"],
+            self.config["vast_password"],
+            storage_base_dir=self.config["vast_storage_base_dir"],
+            api_version=self.config["vast_api_version"],
+            ssl=ssl_verify,
+        )
+
+    async def create_quota_model(self) -> VastQuotaModel:
+        return VastQuotaModel(self.mount_path, self.api_client)
+
+    async def get_capabilities(self) -> FrozenSet[str]:
+        return frozenset(
+            [CAP_VFOLDER, CAP_METRIC, CAP_QUOTA, CAP_FAST_FS_SIZE, CAP_FAST_SCAN, CAP_FAST_SIZE]
```

Comment:
`CAP_FAST_SCAN` capability should only be advertised if the storage supports scanning number of files in a "fast" fashion. In here fast means we should not have to scan the whole filesystem and count the number of files one by one, usually by querying the API vender has exposed. But here I don't see any of such custom methods overridden. Purestorage backend's `RapidFileToolsFSOpModel.scan_tree_usage()` can be a good example of such feature.